pkg/storage/etcd3/watcher.go

```diff
@@ -0,0 +1,330 @@
+			return
+		}
+	}
+}
+
+// transform transforms an event into a result for user if not filtered.
+// TODO (Optimization):
+// - Save remote round-trip.
+//   Currently, DELETE and PUT event don't contain the previous value.
+//   We need to do another Get() in order to get previous object and have logic upon it.
+//   We could potentially do some optimizations:
+//   - For PUT, we can save current and previous objects into the value.
+//   - For DELETE, See https://github.com/coreos/etcd/issues/4620
+func (wc *watchChan) transform(e *event) (res *watch.Event) {
+	curObj, oldObj, err := prepareObjs(wc.ctx, e, wc.watcher.client, wc.watcher.codec, wc.watcher.versioner)
+	if err != nil {
+		go wc.sendError(err)
+		return nil
+	}
+
+	switch {
+	case e.isDeleted:
+		if !wc.filter(oldObj) {
+			return nil
+		}
+		res = &watch.Event{
+			Type:   watch.Deleted,
+			Object: oldObj,
+		}
+	case e.isCreated:
+		if !wc.filter(curObj) {
+			return nil
+		}
+		res = &watch.Event{
+			Type:   watch.Added,
+			Object: curObj,
+		}
+	default:
+		curObjPasses := wc.filter(curObj)
+		oldObjPasses := wc.filter(oldObj)
+		switch {
+		case curObjPasses && oldObjPasses:
+			res = &watch.Event{
+				Type:   watch.Modified,
+				Object: curObj,
+			}
+		case curObjPasses && !oldObjPasses:
+			res = &watch.Event{
+				Type:   watch.Added,
+				Object: curObj,
+			}
+		case !curObjPasses && oldObjPasses:
+			res = &watch.Event{
+				Type:   watch.Deleted,
+				Object: oldObj,
+			}
+		}
+	}
+	return res
+}
+
+func parseError(err error) *watch.Event {
+	var status *unversioned.Status
+	switch {
+	case err == etcdrpc.ErrCompacted:
+		status = &unversioned.Status{
+			Status:  unversioned.StatusFailure,
+			Message: err.Error(),
+			Code:    http.StatusGone,
+			Reason:  unversioned.StatusReasonExpired,
+		}
+	default:
+		status = &unversioned.Status{
+			Status:  unversioned.StatusFailure,
+			Message: err.Error(),
+			Code:    http.StatusInternalServerError,
+			Reason:  unversioned.StatusReasonInternalError,
+		}
+	}
+
+	return &watch.Event{
+		Type:   watch.Error,
+		Object: status,
+	}
+}
+
+func (wc *watchChan) sendError(err error) {
+	select {
+	case wc.errChan <- err:
+	case <-wc.ctx.Done():
+	}
+}
+
+func (wc *watchChan) sendEvent(e *event) {
+	if len(wc.incomingEventChan) == incomingBufferSize {
+		glog.V(2).Infof("Fast watcher, slow processing. Number of buffered events: %d."+
+			"Probably caused by slow decoding, user not receiving fast, or other processing logic",
+			incomingBufferSize)
+	}
+	select {
+	case wc.incomingEventChan <- e:
+	case <-wc.ctx.Done():
+	}
+}
+
+func prepareObjs(ctx context.Context, e *event, client *clientv3.Client, codec runtime.Codec, versioner storage.Versioner) (curObj runtime.Object, oldObj runtime.Object, err error) {
+	curObj, err = decodeObj(codec, versioner, e.value, e.rev)
+	if err != nil {
+		return nil, nil, err
+	}
+	if e.isDeleted || !e.isCreated {
+		getResp, err := client.Get(ctx, e.key, clientv3.WithRev(e.rev-1))
```

Comment:
> I'm wondering about this additional Get here - did you measure the performance impact of it?

I would say do not worry about the perf right now. Let's get this working correctly first. Trust us, we are confident that we will make this faster than today's v2 storage layer together with the etcdv3 improvement and the "optimizations" we are going to make here in future prs.